clang/test/SemaSYCL/intel-fpga-local.cpp

```diff
@@ -874,6 +869,29 @@ void check_template_parameters() {
   //CHECK-NEXT: IntegerLiteral{{.*}}8{{$}}
   [[intel::bank_bits(A, 3), intel::bankwidth(C)]] unsigned int bank_bits_width;
 
+  // Test that checks template instantiations for different arg values.
+  // CHECK: VarDecl{{.*}}bank_bits_args 'unsigned int'
+  // CHECK: IntelFPGAMemoryAttr{{.*}}Implicit Default
+  // CHECK: IntelFPGANumBanksAttr
+  // CHECK-NEXT: ConstantExpr
+  // CHECK-NEXT: value:{{.*}}4
+  // CHECK-NEXT: IntegerLiteral{{.*}}'int' 4
+  // CHECK: IntelFPGANumBanksAttr{{.*}}Implicit
```

Comment:
I assume this is existing behaviors but can you please explain why these attributes are showing up twice?

Reply:
Thanks @elizabethandrews for the reviews and asking this question.

Yes, this is an existing behavior.

The attribute IntelFPGANumBanksAttr is generating twice because of the codes from bank_bit attributes
```
// Check or add the related numbanks attribute.
  if (auto *NBA = D->getAttr<IntelFPGANumBanksAttr>()) {
    Expr *E = NBA->getValue();
    if (!E->isValueDependent()) {
      Expr::EvalResult Result;
      E->EvaluateAsInt(Result, Context);
      llvm::APSInt Value = Result.Val.getInt();
      if (Args.size() != Value.ceilLogBase2()) {
        Diag(TmpAttr.getLoc(), diag::err_bankbits_numbanks_conflicting);
        return;
      }
    }
  } else {
    llvm::APInt Num(32, (unsigned)(1 << Args.size()));
    Expr *NBE =
        IntegerLiteral::Create(Context, Num, Context.IntTy, SourceLocation());
    D->addAttr(IntelFPGANumBanksAttr::CreateImplicit(Context, NBE));
  }
```

```
This part of code  generates one of IntelFPGANumBanksAttr attribute
```} else {
    llvm::APInt Num(32, (unsigned)(1 << Args.size()));
    Expr *NBE =
        IntegerLiteral::Create(Context, Num, Context.IntTy, SourceLocation());
-IntelFPGANumBanksAttr 0x55ad8f126208 <<invalid sloc>>
|     |   | `-ConstantExpr 0x55ad8f126190 <<invalid sloc>> 'int'
|     |   |   |-value: Int 4
|     |   |   `-IntegerLiteral 0x55ad8f11be60 <<invalid sloc>> 'int' 4
```
|     |   |-IntelFPGANumBanksAttr 0x55ad8f126238 <<invalid sloc>> Implicit
|     |   | `-IntegerLiteral 0x55ad8f11be60 <<invalid sloc>> 'int' 4
```
```
This part generates other  IntelFPGANumBanksAttr attribute attr
D->addAttr(IntelFPGANumBanksAttr::CreateImplicit(Context, NBE));
  }

IntelFPGANumBanksAttr 0x55ad8f126238 <<invalid sloc>> Implicit
|     |   | `-IntegerLiteral 0x55ad8f11be60 <<invalid sloc>> 'int' 4
```

Reply:
Sorry if I am misunderstanding something here but the way I see it - 

`Expr *NBE = IntegerLiteral::Create(Context, Num, Context.IntTy, SourceLocation());`

creates an integer literal, which is then attached to implicit attribute here - 

`D->addAttr(IntelFPGANumBanksAttr::CreateImplicit(Context, NBE));`

I don't understand how that generated this AST - 

     // CHECK: IntelFPGANumBanksAttr
     // CHECK-NEXT: ConstantExpr
     // CHECK-NEXT: value:{{.*}}4
     // CHECK-NEXT: IntegerLiteral{{.*}}'int' 4
     // CHECK: IntelFPGANumBanksAttr{{.*}}Implicit
     // CHECK-NEXT: IntegerLiteral{{.*}} 'int' 4

I would have expected the AST similar to AST you added in L615, except with `Implicit`. Something like - 

     // CHECK: IntelFPGABankBitsAttr Implicit
     // CHECK-NEXT: ConstantExpr
     // CHECK-NEXT: value:{{.*}}4
     // CHECK-NEXT: IntegerLiteral{{.*}}4{{$}}

Maybe I'm misunderstanding how AST is generated when there are implicit attributes? @AaronBallman

Reply:
It is dropping "implicit" because of codes:
```
https://github.com/intel/llvm/blob/22be4274cf4e4f09ecfa56fc8fd344109eabb350/clang/lib/Sema/SemaDeclAttr.cpp#L6739-#L6743
```

Reply:
This is full AST 

```
|-DeclStmt 0x561ff7a8dfa0 <line:893:3, col:83>
| |   | `-VarDecl 0x561ff7a8dda0 <col:56, col:69> col:69 bank_bits_args 'unsigned int'
| |   |   |-IntelFPGANumBanksAttr 0x561ff7a8de80 <<invalid sloc>> Implicit
| |   |   | `-IntegerLiteral 0x561ff7a8de60 <<invalid sloc>> 'int' 4
| |   |   |-IntelFPGAMemoryAttr 0x561ff7a8dee0 <<invalid sloc>> Implicit Default
| |   |   `-IntelFPGABankBitsAttr 0x561ff7a8df08 <col:5, col:26>
| |   |     |-DeclRefExpr 0x561ff7a8dd08 <col:22> 'int' NonTypeTemplateParm 0x561ff7a8d430 'A' 'int'
| |   |     `-ConstantExpr 0x561ff7a8de20 <col:25> 'int'
| |   |       |-value: Int 3
| |   |       `-IntegerLiteral 0x561ff7a8dd28 <col:25> 'int' 3

DeclStmt 0x561ff7a983a0 <line:893:3, col:83>
|     | `-VarDecl 0x561ff7a98128 <col:56, col:69> col:69 bank_bits_args 'unsigned int'
|     |   |-IntelFPGAMemoryAttr 0x561ff7a981b0 <<invalid sloc>> Implicit Default
|     |   |-IntelFPGANumBanksAttr 0x561ff7a98208 <<invalid sloc>>
|     |   | `-ConstantExpr 0x561ff7a98190 <<invalid sloc>> 'int'
|     |   |   |-value: Int 4
|     |   |   `-IntegerLiteral 0x561ff7a8de60 <<invalid sloc>> 'int' 4
|     |   |-IntelFPGANumBanksAttr 0x561ff7a98238 <<invalid sloc>> Implicit
|     |   | `-IntegerLiteral 0x561ff7a8de60 <<invalid sloc>> 'int' 4
|     |   |-IntelFPGAMemoryAttr 0x561ff7a98268 <<invalid sloc>> Implicit Default
|     |   |-IntelFPGABankBitsAttr 0x561ff7a98320 <col:5, col:26>
|     |   | |-ConstantExpr 0x561ff7a982e0 <col:22> 'int'
|     |   | | |-value: Int 2
|     |   | | `-SubstNonTypeTemplateParmExpr 0x561ff7a982b0 <col:22> 'int'
|     |   | |   |-NonTypeTemplateParmDecl 0x561ff7a8d430 <line:829:11, col:15> col:15 referenced 'int' depth 0 index 0 A
|     |   | |   `-IntegerLiteral 0x561ff7a98290 <line:893:22> 'int' 2
|     |   | `-ConstantExpr 0x561ff7a98300 <col:25> 'int'
|     |   |   |-value: Int 3
|     |   |   `-IntegerLiteral 0x561ff7a8dd28 <col:25> 'int' 3
|     |   `-IntelFPGABankBitsAttr 0x561ff7a98360 <col:5, col:26>
|     |     |-DeclRefExpr 0x561ff7a8dd08 <col:22> 'int' NonTypeTemplateParm 0x561ff7a8d430 'A' 'int'
|     |     `-ConstantExpr 0x561ff7a8de20 <col:25> 'int'
|     |       |-value: Int 3
|     |       `-IntegerLiteral 0x561ff7a8dd28 <col:25> 'int' 3
````

Reply:
That code is dropping an implicit attribute when there is user specified num_banks attribute. I don' think it should be hit in this case. My concern is from the AST it looks like we have a non-implicit NumBanksAttr and an implicit one. Maybe this is how AST is generated in clang for implicit attributes with a value? I am not sure though, so I will test this and get back to you!

Reply:
FWIW, that sounds like a bug to me. Implicit attributes are not handled any different from non-implicit ones (generally speaking), so I'd have expected to only see the attribute added once.

Reply:
I took a quick look and I THINK it has something to do with template instantiation. After processing `BankBitsAttr` prior to `SemaTemplateInstantiateDecl`.cpp, the AST is as expected - Implicit memory attributes for `IntelFPGANumBanksAttr `and `IntelFPGAMemoryAttr` and user specified `IntelFPGABankBitsAttr `. 
```

VarDecl 0x1038ee50 </localdisk2/eandrews/sandbox/publicsycl/llvm/clang/test/SemaSYCL/intel-fpga-local.cpp:54:56, col:69> col:69 bank_bits_args 'unsigned int'
|-IntelFPGANumBanksAttr 0x1038ef10 <<invalid sloc>> Implicit
| `-IntegerLiteral 0x1038eef0 <<invalid sloc>> 'int' 4
|-IntelFPGAMemoryAttr 0x1038ef70 <<invalid sloc>> Implicit Default
`-IntelFPGABankBitsAttr 0x1038ef98 <col:5, col:26>
  |-DeclRefExpr 0x1038edb8 <col:22> 'int' NonTypeTemplateParm 0x1038e9b0 'A' 'int'
  `-ConstantExpr 0x1038eed0 <col:25> 'int'
    |-value: Int 3
    `-IntegerLiteral 0x1038edd8 <col:25> 'int' 3
```
I haven't debugged this properly but I see during template instantiation, we enter - `AddIntelFPGANumBanksAttr`. I assume this is where the incorrect second attribute is added. I'm not sure why we enter this handler since there is no user `numbanks`.

This is the reproducer I used - 

```
template <int A, int B, int C, int D, int E>
void check_template_parameters() {
  [[intel::bank_bits(A, 3), intel::bank_bits(42, 43)]] unsigned int bank_bits_args;
}
```

I assume we will see this behavior even the duplicate attribute (I haven't verified this)

Reply:
Thanks @AaronBallman and @elizabethandrews for catching this problem. The problem is happening only during template instantiation. I have verified and the same behavior is happening with existing attribute support without this PR.
I have created https://github.com/intel/llvm/issues/5350 to investigate the issue.

Reply:
@smanna12 since you're fixing issues with `bank_bits `attribute in this PR, can you fix it in this PR? 

I suggest running a few tests to see if this issue exists with other attributes as well. If it does, then it might make sense to fix this in a separate patch but otherwise, I think it makes most sense to fix it as part of this patch

Reply:
>>I suggest running a few tests to see if this issue exists with other attributes as well. If it does, then it might make sense to fix this in a separate patch but otherwise, I think it makes most sense to fix it as part of this patch.

Yes, I will check with other attributes. If this happens with bank_bit attribute only, i will fix this as part of this patch. Thanks for the suggestion.